Feature/create user user service

dTarasova opened the pull request:
UserService: CreateUser method
NewsServiceTests project added
NewsService: GetNews output for CB changed

dTarasova committed:
User service - CreateUser, User - constructor added, NewsServiceTests project added

dTarasova committed:
Useless method is deleted

nmamro commented on Services/UserService/Commands/CreateUserCommand.cs:
```diff
@@ -0,0 +1,28 @@
+using Kernel;
+
+public class CreateUserCommand : ICreateUser<string, string>
+
```
Idk there're some strange empty lines in some files

dTarasova replied:
Wow, namespace isn't there. Thanks!

Cerberous007 reviewed (approved):
Overall, everything is good

Maksim-Trolina commented on Integrations/CentralBankIntegrationLib/CentralBankNewsPublisher.cs:
```diff
@@ -20,8 +21,19 @@ public string GetNews()
+                foreach (Valute valute in valutes)
+                {
+                    sb.AppendLine(valute.Description());
+                }
```
Why not use var?After all, the type is specified explicitly

dTarasova replied:
like this? 
var description = valute.Description();
sb.appendline(description? 

It isn't a problem, I can change, but why?

Maksim-Trolina replied:
You must adhere to your chosen code style.

ifelseelif commented on Integrations/CentralBankIntegrationLib/CentralBankNewsPublisher.cs:
```diff
@@ -20,8 +21,19 @@ public string GetNews()
             var sb = new StringBuilder();
-            sb.AppendLine("Top - 5 exchange rates:");
-            sb.AppendLine(valutes != null ? string.Join('\n', valutes) : Errors.CanNotGetCurrencyRateFromCB);
+            // sb.AppendLine(valutes != null ? string.Join('\n', valutes) : Errors.CanNotGetCurrencyRateFromCB);
```
What is this comment for?

dTarasova replied:
I forgot that git saves all versions and I don't need to have this :)
It was shorter way to write this function but now it returnes no correct data

dTarasova committed:
Deleted useless comments, namespace added

DanilFadeeev commented on Common/DTO/User.cs:
```diff
@@ -2,6 +2,11 @@
+        {
+            Email = email;
+            Password = password;
+        }
```
may be need empity constructor too? I think, this need to correct work <Editform> tag in razor

dTarasova replied:
I can add it. But I think it had to be added by people who need this. Are your command need this right now? I hadn't worked with razor pages yet.
Also, there was an idea to create class UserEmailPassword (don't know how to name it) With only 2 properties: email and password, obviously. It could be useful to send less data

DanilFadeeev replied:
may be create email_id_user and id_login_user and oter? Why we need create another DTO? Ok, how we can set user ID and login? it is readonly

dTarasova replied:
Okay, probably just id, email, password. Look, User will have more properties like LastName, FirstName, Birthdate, etc. And I think it isn't good to move all structure when we just want to register or authorize.

dTarasova replied:
not actual task

DanilFadeeev commented on Services/UserService/Interfaces/ICreateCommand.cs:
```diff
@@ -0,0 +1,7 @@
+    {
+        public Output Execute(Input email, Input password);
+    }
+}
```
why we use generic to create conkrect object user?

studokim replied:
Do we need such interface at all? Why not ICommand?

dTarasova replied:
Yes, I think . We still have positive elements in using interface. Spartak said that both of ways are okay. I don't sure if we wouldn't have the same signature in Create / Edit / Delete User. I want to be sure that my code wouldn't fail anything else. If it would become like we have different arguments in all actions we change it to ICommand

DarkFelas replied:
And a lot of code with it, all our 100 and more unit tests, injections, signatures in all methods. Yes really good idea to change it latter ;-)

dTarasova committed:
Merge branch 'develop' into feature/createUserUserService

dTarasova committed:
Some errors fixed

studokim commented on Integrations/CentralBankIntegrationLib/CentralBankNewsPublisher.cs:
```diff
@@ -20,8 +19,19 @@ public string GetNews()
-            sb.AppendLine(valutes != null ? string.Join('\n', valutes) : Errors.CanNotGetCurrencyRateFromCB);
+            if (valutes.Count == 0)
+            {
+                sb.AppendLine("Valutes are not founded");
```
Why Valute, not Currency?
"founded" must be replaced with "found".
Or even the whole string: "No currencies found".

dTarasova replied:
Ok

studokim commented on Integrations/CentralBankIntegrationLib/CentralBankNewsPublisher.cs:
```diff
@@ -20,8 +19,19 @@ public string GetNews()
+                foreach (Valute valute in valutes)
+                {
+                    var description = valute.Description();
+                    sb.AppendLine(description);
```
I suppose the _Why not use var_ comment by Maksim wasn't about this. 
`sb.AppendLine(valute.Description());` must be okay, meanwhile `foreach (var valute in valutes)` better fits out convention

dTarasova replied:
fixed

studokim commented on Integrations/CentralBankIntegrationLib/XmlSerializationObjects/ValCurse.cs:
```diff
@@ -13,6 +13,6 @@ public class ValCurs
         public List<Valute> Valutes { get; set; }
 
-        public List<string> TopValutesCharCode => new List<string> { "EUR", "USD", "GBP", "CNY", "JPY" };
+        public List<string> TopValutesCharCode => new List<string> {  };
```
Should we initialize the property at all? If so, then why with =>?

dTarasova replied:
Right now it is the way to define top currencies. Sorry, don't understand what do you suggest?

dTarasova committed:
CentralBank integration fixed

DarkFelas commented on Common/DTO/User.cs:
```diff
@@ -2,6 +2,11 @@
+            Email = email;
+            Password = password;
+        }
         public int Id { get; }
```
How we will set id if we need it?

DarkFelas commented on Services/UserService/Startup.cs:
```diff
@@ -22,14 +22,15 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+
             app.UseRouting();
 
+            app.UseAuthorization();
```
For what?

DarkFelas commented on Services/UserService/Commands/CreateUserCommand.cs:
```diff
@@ -0,0 +1,32 @@
+            return message;
+        }
+        private string createUserInDatabase(string email, string password)
+        {
```
Not sure we need InDatabase

DarkFelas commented on Integrations/CentralBankIntegrationLib/CentralBankNewsPublisher.cs:
```diff
@@ -20,8 +19,19 @@ public string GetNews()
             var sb = new StringBuilder();
-            sb.AppendLine("Top - 5 exchange rates:");
-            sb.AppendLine(valutes != null ? string.Join('\n', valutes) : Errors.CanNotGetCurrencyRateFromCB);
+            if (valutes.Count == 0)
```
What will be if valutes == null ?

DarkFelas commented on Services/NewsService/Controllers/NewsPublisherController.cs:
```diff
@@ -11,10 +11,15 @@ namespace NewsService.Controllers
+        /// </summary>
         public String GetNews([FromQuery] NewsPublisherTypes newsPublisherType)
         {
             INewsPublisher newsPublisher = NewsPublisherFactory.CreateNewsPublisher(newsPublisherType);
```
Not sure we need that name for Create method, we already have NewsPublisherFactory as parent object, for what this duplication?

DarkFelas commented on Services/UserService/Commands/CreateUserCommand.cs:
```diff
@@ -0,0 +1,32 @@
+namespace UserService.Commands
+{
+    public class CreateUserCommand : ICreateUser<string, string>
+
```
remove empty line

DarkFelas commented on Services/UserService/Controllers/UserServiceController.cs:
```diff
@@ -0,0 +1,23 @@
+        /// </summary>
+        public string CreateUser([FromQuery] string email, string password)
+        {
+            ICreateUser<string, string> command = new CreateUserCommand();
```
Use injection for command object

DarkFelas commented on Services/UserService/Controllers/UserServiceController.cs:
```diff
@@ -0,0 +1,23 @@
+        /// <summary>
+        /// This method will be implemented in communication with other services
+        /// </summary>
+        public string CreateUser([FromQuery] string email, string password)
```
Every argument should have attribute

DarkFelas commented on Services/UserService/Interfaces/ICreateCommand.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public interface ICreateUser<in Input, out Output>
+    {
+        public Output Execute(Input email, Input password);
```
public ?

DarkFelas commented on Tests/NewsServiceTests/Controllers/GetNewsTests.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace NewsServiceTests.Controllers
+{
+    class GetNewsTests
```
Bad name, i do not understand what object is a target for tests

DarkFelas reviewed (changes requested):
Need more work

dTarasova committed:
UserService: injection fixed, Interface of command deleted
DTO.User: Constructor changed,  type of Id changed to Guid

dTarasova committed:
CentralBankIntegrationLib null reference exception catched
NewsServiceController refactored
DTO.User constructor changed
UserService injection in comands

dTarasova committed:
NewsServiceTests: GetNewsTests deleted

dTarasova committed:
Merge branch 'develop' into feature/createUserUserService

# Conflicts:
#	TPlatform.sln

DarkFelas commented on Integrations/CentralBankIntegrationLib/CentralBankNewsPublisher.cs:
```diff
@@ -19,19 +20,27 @@ public string GetNews()
 
             var sb = new StringBuilder();
-            if (valutes.Count == 0)
+            try
```
do not use try catch in such case, it is not a solution for such small problem.
you can just check valutes != null

dTarasova replied:
fixed

DarkFelas commented on Services/UserService/Commands/CreateUserCommand.cs:
```diff
@@ -1,10 +1,10 @@
 
 namespace UserService.Commands
 {
-    public class CreateUserCommand : ICreateUser<string, string>
```
why you remove interface, we need them if we want create good unit tests

dTarasova replied:
Why ? What interfaces can make to improve unit tests? 
I thought it was useless, if I user only ICreateCommand

dTarasova replied:
fixed

DarkFelas commented on Services/UserService/Commands/CreateUserCommand.cs:
```diff
@@ -13,16 +13,16 @@ public string Execute(string email, string password)
             {
-                var user = new User(email, password);
-                message = createUserInDatabase(email, password);
+                var user = new User(email, password, Guid.Empty);
```
Guid.Empty wtf????

dTarasova replied:
It wouldn't be necessary if we create new model UserCredential
UserService can't create ID , but it had to be set in constructor. That was the way we agreed in chat in thursday. But it will be removed

dTarasova replied:
This code was deleted

DarkFelas commented on Services/UserService/Controllers/UserServiceController.cs:
```diff
@@ -1,22 +1,20 @@
         /// This method will be implemented in communication with other services
         /// </summary>
-        public string CreateUser([FromQuery] string email, string password)
+        public string CreateUser([FromServices] CreateUserCommand command, [FromBody] string email, [FromBody] string password)
```
you can not have 2 body in one request, rework

dTarasova replied:
fixed

DarkFelas commented on Services/UserService/Startup.cs:
```diff
@@ -2,7 +2,7 @@
 using Microsoft.Extensions.Hosting;
-
+using UserService.Commands;
 namespace UserService
```
add blank line between using block and namespace

dTarasova replied:
fixed

DarkFelas reviewed (changes requested):
i do not understand why in UserService feature we have some NewsService changes?

dTarasova committed:
UserCredentials added, CreateUserCommand refactored, User fixed

dTarasova committed:
EmailValidation Exception created

dTarasova committed:
Answer for create user changed to HttpStatusCode

dTarasova committed:
Controller name fixed

dTarasova committed:
Merge branch 'develop' into feature/createUserUserService

# Conflicts:
#	Services/UserService/Controllers/UserController.cs
#	Services/UserService/Startup.cs
#	Services/UserService/UserService.csproj
#	TPlatform.sln

dTarasova committed:
small refactoring

nmamro commented on Common/Kernel/CommonValidations.cs:
```diff
@@ -13,5 +13,18 @@ public static void ValidateId(int id)
+        {
+
+            var addr = new MailAddress(email); 
+            bool emailIsValid = addr.Address == email;
```
Wow wow wow, do we really need this variable???

dTarasova replied:
fixed

nmamro commented on Common/Kernel/CommonValidations.cs:
```diff
@@ -13,5 +13,18 @@ public static void ValidateId(int id)
+            
+            if (emailIsValid == false)
+            {
+                throw new Exception(Errors.EmailCanNotBeNullEmptyOrNotEmailFormat);
```
May be we should use another type of Exception? For example, a custom one.

dTarasova replied:
It is a custom exception that I created

nmamro replied:
I mean a custom exception as a class, inheriting `Exception`, not a custom error message

DarkFelas replied:
It is not a custom exception, Nikita is right.

nmamro commented on Services/NewsService/Controllers/NewsPublisherController.cs:
```diff
@@ -11,14 +10,18 @@ namespace NewsService.Controllers
         [HttpGet]
+
+        ///<summary>
+        /// Return news depends on a publisher type
```
_depending_?

dTarasova replied:
fixed

nmamro commented on Services/UserService/Commands/CreateUserCommand.cs:
```diff
@@ -0,0 +1,23 @@
+            CommonValidations.ValidateEmail(userCredential.Email);
+            return createUser(userCredential);
+        }
+        private HttpStatusCode createUser(UserCredential userCredential)
```
Missing empty line between methods

dTarasova replied:
fixed

dTarasova committed:
TPlatform.sln changed

dTarasova committed:
Delete useless variables and add empty lines between methods

DarkFelas commented on Common/DTO/User.cs:
```diff
@@ -1,8 +1,10 @@
+        public Guid Id { get; }
         public string Login { get; }
         public string Email { get; set; }
         public string Password { get; }
```
Still just Password not PasswordHash?

dTarasova replied:
Fixed

DarkFelas commented on Common/DTO/UserCredential.cs:
```diff
@@ -0,0 +1,16 @@
+            Password = password;
+        }
+        public string Email { get; set; }
+        public string Password { get; }
```
I don't understand why password here and password in User model?

dTarasova replied:
Why is it bad?

dTarasova replied:
Edit function can edit password and other data. But we have to transmit the same object - user. So I think that logic would be the same if we edit password and name, for example.

DarkFelas commented on Common/DTO/UserCredential.cs:
```diff
@@ -0,0 +1,16 @@
+    /// </summary>
+    public class UserCredential
+    {
+        public UserCredential(string email, string password)
```
It is a good practice to put class methods after class fields and class properties

dTarasova replied:
fixed

DarkFelas commented on Common/Kernel/CommonValidations.cs:
```diff
@@ -13,5 +13,16 @@ public static void ValidateId(int id)
+        public static void ValidateEmail(string email)
+        {
+
+            var addr = new MailAddress(email);             
```
Without try catch this will throw unhandled exception and whole system will be down

dTarasova replied:
fixed

DarkFelas commented on Common/Kernel/Properties/Errors.resx:
```diff
@@ -120,6 +120,9 @@
     <value>Can't get currency rate from Central Bank.</value>
   </data>
+  <data name="EmailCanNotBeNullEmptyOrNotEmailFormat" xml:space="preserve">
+    <value>Email is null, empty or not match email address format</value>
```
Put dot (.) in the end

dTarasova replied:
fixed

DarkFelas commented on Services/NewsService/Controllers/NewsPublisherController.cs:
```diff
@@ -11,14 +10,18 @@ namespace NewsService.Controllers
         {
-            INewsPublisher newsPublisher = NewsPublisherFactory.CreateNewsPublisher(newsPublisherType);
-            return newsPublisher.GetNews();
+            return  NewsPublisherFactory.CreateNewsPublisher(newsPublisherType).GetNews();
```
Replace CreateNewsPublisher() with Create()

dTarasova replied:
fixed

dTarasova replied:
Fixed

DarkFelas commented on Services/UserService/Commands/CreateUserCommand.cs:
```diff
@@ -0,0 +1,24 @@
+    public class CreateUserCommand : ICreateUserCommand
+
+    {
+        public HttpStatusCode Execute(UserCredential userCredential)
```
Do not return HttpStatusCode it is a bad practice

dTarasova replied:
Fixed

DarkFelas commented on Services/UserService/Commands/CreateUserCommand.cs:
```diff
@@ -0,0 +1,24 @@
+            return createUser(userCredential);
+        }
+
+        private HttpStatusCode createUser(UserCredential userCredential)
```
HttpStatusCode

dTarasova replied:
fixed

DarkFelas commented on Services/UserService/Controllers/UserController.cs:
```diff
@@ -1,14 +1,27 @@
+        /// <summary>
+        /// This method will be implemented in communication with other services
+        /// </summary>
+        public HttpStatusCode CreateUser([FromServices] ICreateUserCommand command, [FromBody] UserCredential userCredential)
```
HttpStatusCode == bad

dTarasova replied:
fixed

DarkFelas commented on Services/UserService/Interfaces/ICreateUserCommand.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public interface ICreateUserCommand
+    {
+        HttpStatusCode Execute(UserCredential userCredential);
```
HttpStatusCode

dTarasova replied:
fixed

DarkFelas reviewed (changes requested):
need more work

dTarasova committed:
NewsService: change name of the method

dTarasova committed:
DTO: User and UserCredential - change name of the properties, swap constructor

dTarasova committed:
UserService CreateUser: type of returned valu changed, exception added

dTarasova committed:
CommonValidation: EmailValidation try - catch added

dTarasova committed:
Errors: dot added

dTarasova committed:
Merge branch 'develop' into feature/createUserUserService

dTarasova committed:
UserService : UserCredential changed to UserEmailPassword

dTarasova committed:
Merge branch 'develop' into feature/createUserUserService

DanilFadeeev commented on Services/UserService/Commands/CreateUserCommand.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace UserService.Commands
+{
+    public class CreateUserCommand : ICreateUserCommand
```
may be rename interface ICreateUserCommand and IDeleUserComand to Comand? They contains identic method Execute

dTarasova replied:
They will have similar Input/output signature. So, it is only the first method will be implemented always
